libs/books/data-access/src/lib/+state/reading-list.effects.spec.ts

```diff
@@ -42,4 +60,92 @@ describe('ToReadEffects', () => {
       httpMock.expectOne('/api/reading-list').flush([]);
     });
   });
+  
+  describe('addBook$', () => {
+    it('should add a book to the reading list successfully', fakeAsync(() => {
+      actions.next(ReadingListActions.addToReadingList({ book, add: false  }));
+
+      effects.addBook$.subscribe(action => {
+        expect(action).toEqual(
+          ReadingListActions.confirmedAddToReadingList({ book, add: false })
+        );
+      })
+
+      httpMock.expectOne('/api/reading-list').flush({});
+    }));
+
+
+    it('should undo the added book when API returns error', fakeAsync(() => {
+      actions.next(ReadingListActions.addToReadingList({ book, add: false}));
+
+      effects.addBook$.subscribe(action => {
+        expect(action).toEqual(
+          ReadingListActions.failedAddToReadingList({ book })
+        );
+      });
+
+      httpMock.expectOne('/api/reading-list').flush({}, { status: 500, statusText: 'server error' });
+
+    }));
+
+    it('show snackbar on add book', async() => {
+      actions = new ReplaySubject();
+      actions.next(ReadingListActions.confirmedAddToReadingList({ book, add:false }));
+
+      effects.showSnackBarOnAdd$.subscribe(() => {
+        snackbar.open(`Added ${book.title} to reading list`,'Undo', { duration: 2000}).onAction().subscribe((action)=>{
+          expect(action).toEqual(
+            ReadingListActions.removeFromReadingList({ item:bookItem, remove:true })
+        )
+        })
+      });
+    });
+
+  });  
+
+  describe('removeBook$', () => {
+    it('should remove book successfully from reading list', done => {
+      actions.next(ReadingListActions.removeFromReadingList({ item, remove: false }));
+
+      effects.removeBook$.subscribe(action => {
+        expect(action).toEqual(
+          ReadingListActions.confirmedRemoveFromReadingList({ item, remove: false  })
+        );
+
+        done();
+      });
+
+      httpMock.expectOne(`/api/reading-list/${item.bookId}`).flush({});
+    });
+
+
+    it('should undo removed book when API returns error', fakeAsync(() => {
+      actions.next(ReadingListActions.removeFromReadingList({ item, remove: false  }));
+
+      effects.removeBook$.subscribe(action => {
+        expect(action).toEqual(
+          ReadingListActions.failedRemoveFromReadingList({ item })
+        );
+      });
+
+      httpMock.expectOne(`/api/reading-list/${item.bookId}`).flush({}, { status: 500, statusText: 'server error' });
+    }));
+
+    it('show snackbar on remove book', async() => {
+      actions = new ReplaySubject();
+      actions.next(ReadingListActions.confirmedRemoveFromReadingList({ item, remove:false }));
+
+      effects.showSnackBarOnRemove$.subscribe(() => {
```

Comment:
This observable should be unsubscribed from when finished

Reply:
Resolved